10-habits/progress/internal/server/server.go

```diff
@@ -33,47 +37,81 @@ type repository interface {
 }
 
 // New returns a Server that can Listen.
-func New(repo repository) *Server {
+func New(interceptorOutput io.Writer, repo Repository) *Server {
 	return &Server{
-		db: repo,
+		interceptorOutput: interceptorOutput,
+		db:                repo,
 	}
 }
 
 // Listen starts the listening to the port.
 func (s *Server) Listen(ctx context.Context, port int) error {
```

Comment:
Here's how to handle all the startup & shutdown issues. I didn't try running the code, but it should be roughly correct.

```
type Config struct {
	Listen      string // default should be 127.0.0.1:8080 or similar
	DebugListen string // default should be 127.0.0.1:6060 or similar
}

// Server is the implementation of the gRPC server.
type Server struct {
	api.UnimplementedHabitsServer
	interceptorOutput io.Writer

	config Config
	log    Log
	db     Repository
}

type Repository interface {
	Add(ctx context.Context, habit habit.Habit) error
	FindAll(ctx context.Context) ([]habit.Habit, error)
}

// New returns a Server that can Listen.
func New(log, config Config, interceptorOutput io.Writer, repo Repository) *Server {
	return &Server{
		interceptorOutput: interceptorOutput,
		db:                repo,
		log:               log,
		config:            config,
	}
}


// Listen starts the listening to the port.
func (s *Server) Listen(ctx context.Context, port int) (err error) {
	group, ctx := errgroup.WithContext(ctx)
	defer func() {
		err = group.Wait()
		if errors.Is(err, context.Canceled) {
			err = nil
		}
	}()

	// SIDENOTE: this uses listenconfig so it can call Listen with the context, meaning you won't need a separate
	// logic for handling ctx finishing. It's still possible to use `group.Go(` approach by adding a separate ctx wait.
	// the alternative is shown below.
	listener, err := net.ListenConfig{}.Listen(ctx, "tcp", s.config.Listen)
	if err != nil {
		return fmt.Errorf("unable to listen on %v: %w", s.config.Listen, err)
	}

	grpcServer := s.registerGRPCServer()
	s.log.Infof("unable to listen on %v", s.config.Listen)

	group.Go(func() error {
		err := grpcServer.Serve(listener)
		errStop := grpcServer.GracefulStop() // not sure whether this is necessary
		return errors.Join(err, errStop)
	})

	if s.config.DebugListen != "" {
		group.Go(func() error {
			listener, err := net.ListenConfig{}.Listen(ctx, "tcp", s.config.DebugListen)
			if err != nil {
				s.log.Infof("error while serving pprof: %v", err)
				return nil
			}

			// SIDENOTE: this approach avoids collision on the global handler
			mux := http.NewServeMux()
			mux.HandleFunc("/debug/pprof/", pprof.Index)
			mux.HandleFunc("/debug/pprof/cmdline", pprof.Cmdline)
			mux.HandleFunc("/debug/pprof/profile", pprof.Profile)
			mux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
			mux.HandleFunc("/debug/pprof/trace", pprof.Trace)

			server := &Server{
				Addr: s.config.DebugListen,
				Handler: mux,
			}
			err := server.Serve(listener)
			if errors.Is(err, context.Canceled) {
				err = nil
			}
			if err != nil {
				s.log.Infof("error while serving pprof: %v", err)
			}

			// ignore pprof starting errors
			return nil
		})
	}
	
	return group.Wait()
}
```

and the alternative without using listenconfig:

```
	group.Go(func() error {
		return grpcServer.Serve(listener)
	})
	group.Go(func() error {
		<-ctx.Done()
		return grpcServer.GracefulStop()
	})

	// and for debug
	if s.config.DebugListen != "" {
		server := &Server{ ... }
		group.Go(func() error {
			err := server.ListenAndServe()
			// log error
			return nil
		}
		group.Go(func() error {
			<-ctx.Done()
			err := server.Shutdown(context.WithTimeout(context.Background(), 10 * time.Second))
			// log error
			return nil
		}
```